physical/redis.go

```diff
@@ -0,0 +1,370 @@
+	conn := c.pool.Get()
+	defer conn.Close()
+
+	// Construct a "directory"-style name with a trailing slash from
+	// `prefix`.  But`prefix` may be the empty string, so be careful
+	// how we add "/".
+	realPrefix := c.keyPath(prefix)
+	if !strings.HasSuffix(realPrefix, "/") {
+		realPrefix += "/"
+	}
+
+	// Ask Redis to list all keys beneath our prefix.  Note that this
+	// is not terribly efficient if you have a lot of keys, but we hope
+	// it's not a common operating.  There are more complex APIs for
+	// doing this incrementally.
+	reply, err := conn.Do("KEYS", realPrefix+"*")
+	matches, err := redis.Strings(reply, err)
+
+	// The KEYS command recurses "subdirectories", but we don't want to
+	// return the recursive values in our output.  This means stripping
+	// everything after "/" and removing duplicates.
+	results := make([]string, 0)
+	dirs := make(map[string]bool)
+	for _, match := range matches {
+		// Remove the path that we were querying from the results.
+		match = strings.TrimPrefix(match, realPrefix)
+
+		slashPos := strings.Index(match, "/")
+		if slashPos == -1 {
+			// We have an ordinary file, so return it directly.
+			results = append(results, match)
+		} else {
+			// Keep the trailing slash.
+			dir := match[0 : slashPos+1]
+
+			// Have we seen this directory already?  If not,
+			// return it and mark it as seen.
+			if _, present := dirs[dir]; !present {
+				results = append(results, dir)
+				dirs[dir] = true
+			}
+		}
+	}
+
+	return results, err
+}
+
+func (c *RedisBackend) LockWith(key, value string) (Lock, error) {
+	return &RedisLock{
+		key:      fmt.Sprintf("%s/_lock/%s", c.path, key),
+		value:    value,
+		pool:     c.pool,
+		logger:   c.logger,
+		lockConf: c.lockConf,
+	}, nil
+}
+
+// A lock implemented using Redis.  This is based on the single-Redis
+// example code used to explain Redlock at http://redis.io/topics/distlock
+// but it might be useful to upgrade this to a full Redlock implementation.
+type RedisLock struct {
+	key      string
+	value    string
+	pool     *redis.Pool
+	logger   *log.Logger
+	lockConf *RedisLockConfig
+}
+
+// Acquire the lock.  To interrupt lock acquistion, close stopCh.  The
+// returned channel will be closed if the lock is lost.
+func (c *RedisLock) Lock(stopCh <-chan struct{}) (<-chan struct{}, error) {
+	// Round up leaderTTL to the nearest second and convert to a
+	// string.  We do this because EXPIRE only has 1-second accuracy.
+	ttlSeconds := float64(c.lockConf.leaderTTL) / float64(time.Second);
+	ttlStr := strconv.Itoa(int(math.Ceil(ttlSeconds)));
+
+	// Loop until we get a non-nil reply from Redis indicating a
+	// successful operation.
+	var reply interface{}
+	for reply == nil {
+		conn := c.pool.Get()
+		defer conn.Close()
+
+		// Attempt to set our key.  "NX" means to only set the key
+		// if it does not exist, and "EX" specifies a timeout in
+		// seconds.
+		reply, err := conn.Do("SET", c.key, c.value, "NX", "EX", ttlStr)
```

Comment:
@emk you can go back to using `PX` and then you don't have to convert to seconds